File: core/x/javaapi/api/src/main/java/viaduct/java/api/reflect/Field.java
package viaduct.java.api.reflect;

import java.util.Objects;
import viaduct.java.api.types.GRT;

/**
 * Describes static properties of a GraphQL field.
 *
 * @param <P> the GRT on which the field is defined
 */
public interface Field<P extends GRT> {

  /** Returns the GraphQL name of this field. */
  String getName();

  /** Returns the descriptor of the type on which this field is defined. */
  Type<P> getContainingType();

  /** Creates a descriptor for a scalar field. */
  static <P extends GRT> Field<P> of(String name, Type<P> containingType) {
    return new FieldDescriptor<>(name, containingType);
  }
}

record FieldDescriptor<P extends GRT>(String name, Type<P> containingType) implements Field<P> {
  FieldDescriptor {
    Objects.requireNonNull(name, "name");
    Objects.requireNonNull(containingType, "containingType");
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public Type<P> getContainingType() {
    return containingType;
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/reflect/Field.kt
package viaduct.api.reflect

import viaduct.api.types.FieldPresenceProbe
import viaduct.api.types.GRT
import viaduct.api.types.InputLike
import viaduct.apiannotations.StableApi

/** A Field describes static properties of a GraphQL field */
@StableApi
interface Field<Parent : GRT> {
    /** the GraphQL name of this field */
    val name: String

    /** the descriptor of the type that this field is mounted on */
    val containingType: Type<Parent>
}

/**
 * Returns whether this input contains a value for [field] after GraphQL defaults are applied.
 * Explicit `null` counts as present. An omitted field with a schema default is present; an omitted
 * field without a default is absent.
 *
 * Defined as an extension on [InputLike] ([Input][viaduct.api.types.Input] /
 * [Arguments][viaduct.api.types.Arguments]) so it is discoverable via completion on an input
 * instance and cannot be called on output-object types.
 */
@StableApi
fun <Parent : InputLike> Parent.isPresent(field: Field<Parent>): Boolean = (this as? FieldPresenceProbe)?.isFieldPresent(field.name) ?: false
